examples/ShowPlayer.cpp

```diff
@@ -270,8 +273,26 @@ void ShowPlayer::SendEntry(const ShowEntry &entry) {
   SendFrame(entry);
   m_playback_pos += entry.next_wait;
 
+  unsigned int timeout = entry.next_wait;
+  if (!m_simulate) {
+    // Using int64_t for target_delta here because
+    // we have to lose 1 bit anyway as InMilliSeconds() returns a signed
+    // 64-bit integer.
+    ola::TimeStamp now;
+    m_clock.CurrentMonotonicTime(&now);
+    int64_t target_delta = m_playback_pos - m_start_playback_pos;
+    int64_t current_delta = (now - m_start_ts).InMilliSeconds();
+    int64_t delay = target_delta - current_delta;
+    if (delay < 0) {
+      OLA_WARN << "Frame at line " << m_loader.GetCurrentLineNumber()
+               << " was meant to have completed " << -delay << " ms ago."
+               << " System too slow?";
+      delay = 0;
+    }
+    timeout = delay;
```

Comment:
The `delay` value is already the actual timeout value (which may be negative if the system is too slow to keep up with the frames in the recording, or positive if the player needs to delay for the next frame).

The idealized simulated one comes from `unsigned int timeout = entry.next_wait;`, which will be overwritten in the `!m_simulate` branch with the `delay` value.

Unless you mean to have the idealized calculation inside the branch instead?

Reply:
I was meaning something like this:
```
  int64_t delay = entry.next_wait;
  if (!m_simulate) {
    // Using int64_t for target_delta here because we have to lose 1 bit anyway
    // as InMilliSeconds() returns a signed 64-bit integer.
    ola::TimeStamp now;
    m_clock.CurrentMonotonicTime(&now);
    int64_t target_delta = m_playback_pos - m_start_playback_pos;
    int64_t current_delta = (now - m_start_ts).InMilliSeconds();
    delay = target_delta - current_delta;
    if (delay < 0) {
      OLA_WARN << "Frame at line " << m_loader.GetCurrentLineNumber()
               << " was meant to have completed " << -delay << " ms ago."
               << " System too slow?";
      delay = 0;
    }

#if UINT_MAX < INT64_MAX
    if (delay > UINT_MAX) {
      OLA_WARN << "Calculated delay of " << delay << " ms"
               << " exceeded maximum delay of " << UINT_MAX << " ms."
               << " Clamping to maximum.";
      delay = UINT_MAX;
    }
#endif
  }
  // Set when next to send data
  RegisterNextTimeout(delay);
```

Unless you were going to say add some debug level stats about how delay and ideal timeout compare each time, rather than just when it won't fit, then tracking them both independently seemed a bit irrelevant.

Given they're essentially equivalent. Or you could possibly even do something like this (i.e. the non-simulate run corrects the idealised view):
```
  int64_t timeout = m_playback_pos - m_start_playback_pos;
  if (!m_simulate) {
    // Get now
    int64_t current_delta = (now - m_start_ts).InMilliSeconds();
    timeout = timeout - current_delta;
    if (delay < 0) {
      OLA_WARN << "Frame at line " << m_loader.GetCurrentLineNumber()
               << " was meant to have completed " << -delay << " ms ago."
               << " System too slow?";
      delay = 0;
    }
    // Clamping
  }
  // Set when next to send data
  RegisterNextTimeout(delay);
```

Reply:
Hm, `m_start_ts` actually starts tracking from playback start, and it doesn't track from the end of the previous frame.

That's why I needed to evaluate both `target` and `current`, since I need to know how far off we are from the idealized time we want to change the timeout.

If we subtract timeout by `current_delta` we'd most likely get a negative value :laughing:

Reply:
Yes, but m_playback_pos is where we should be when we've finished this delay (in relative time) and m_start_playback_pos is where in the showfile we started (relative).

So `m_playback_pos - m_start_playback_pos` should be the difference using maths from relative points and `now - m_start_ts` should be the actual difference using clock time.

> If we subtract timeout by `current_delta` we'd most likely get a negative value

I don't follow, the maths is the same as your PR currently does, it's just I've reduced the number of variables and slightly tweaked how its laid out.

> Given they're essentially equivalent. Or you could possibly even do something like this (i.e. the non-simulate run corrects the idealised view):

Looks like I made a typo before, it still needs a little bit of work but I think something along those lines should be possible.

Reply:
If you still remember how this works I think you could make that change?